lib/backend/k8s/resources/ipam_affinity.go

```diff
@@ -0,0 +1,269 @@
+)
+
+func NewBlockAffinityClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sResourceClient {
+	// Create a resource client which manages k8s CRDs.
+	rc := customK8sResourceClient{
+		clientSet:       c,
+		restClient:      r,
+		name:            BlockAffinityCRDName,
+		resource:        BlockAffinityResourceName,
+		description:     "Calico IPAM block affinities",
+		k8sResourceType: reflect.TypeOf(apiv3.BlockAffinity{}),
+		k8sResourceTypeMeta: metav1.TypeMeta{
+			Kind:       apiv3.KindBlockAffinity,
+			APIVersion: apiv3.GroupVersionCurrent,
+		},
+		k8sListType:  reflect.TypeOf(apiv3.BlockAffinityList{}),
+		resourceKind: apiv3.KindBlockAffinity,
+	}
+
+	return &affinityBlockClient{rc: rc}
+}
+
+// affinityBlockClient implements the api.Client interface for BlockAffinity objects. It
+// handles the translation between v1 objects understood by the IPAM codebase in lib/ipam,
+// and the CRDs which are used to actually store the data in the Kubernetes API.
+// It uses a customK8sResourceClient under the covers to perform CRUD operations on
+// kubernetes CRDs.
+type affinityBlockClient struct {
+	rc customK8sResourceClient
+}
+
+// toV1 converts the given v3 CRD KVPair into a v1 model representation
+// which can be passed to the IPAM code.
+func (c affinityBlockClient) toV1(kvpv3 *model.KVPair) (*model.KVPair, error) {
+	// Parse the CIDR into a struct.
+	_, cidr, err := net.ParseCIDR(kvpv3.Value.(*apiv3.BlockAffinity).Spec.CIDR)
+	if err != nil {
+		log.WithField("cidr", cidr).WithError(err).Error("failed to parse cidr")
+		return nil, err
+	}
+	state := model.BlockAffinityState(kvpv3.Value.(*apiv3.BlockAffinity).Spec.State)
+	return &model.KVPair{
+		Key: model.BlockAffinityKey{
+			CIDR: *cidr,
+			Host: kvpv3.Value.(*apiv3.BlockAffinity).Spec.Node,
+		},
+		Value: &model.BlockAffinity{
+			State: state,
+		},
+		Revision: kvpv3.Revision,
+		UID:      &kvpv3.Value.(*apiv3.BlockAffinity).UID,
+	}, nil
+}
+
+// parseKey parses the given model.Key, returning a suitable name, CIDR
+// and host for use in the Kubernetes API.
+func (c affinityBlockClient) parseKey(k model.Key) (name, cidr, host string) {
+	host = k.(model.BlockAffinityKey).Host
+	cidr = fmt.Sprintf("%s", k.(model.BlockAffinityKey).CIDR)
+	cidrname := names.CIDRToName(k.(model.BlockAffinityKey).CIDR)
+
+	// Include the hostname as well.
+	host = k.(model.BlockAffinityKey).Host
+	name = fmt.Sprintf("%s-%s", host, cidrname)
+
+	if len(name) >= 253 {
+		// If the name is too long, we need to shorten it.
+		// Remove enough characters to get it below the 253 character limit,
+		// as well as 11 characters to add a hash which helps with uniqueness,
+		// and two characters for the `-` separators between clauses.
+		name = fmt.Sprintf("%s-%s", host[:252-len(cidrname)-13], cidrname)
+
+		// Add a hash to help with uniqueness.
+		// Kubernetes requires all names to end with an alphabetic character, so
+		// append a 'c' to the end to ensure we always meet this requirement.
```

Comment:
Where does this req come from?  I'm sure I've created pods with names like `busybox-1` before, would be nice to avoid having to add an extra random char to the end if we can

Reply:
Yeah, good spot. That was something from the original prototype which may have been misguided.